tiledb/libtiledb.pyx

```diff
@@ -2991,57 +3052,64 @@ cdef class Array(object):
             _raise_ctx_err(ctx_ptr, rc)
         return
 
-    def __init__(self, uri, mode='r', key=None, timestamp=None, attr=None, Ctx ctx=default_ctx()):
-        # ctx
+    @staticmethod
+    def load_typed(uri, mode='r', key=None, timestamp=None, attr=None, Ctx ctx=default_ctx()):
+        cdef int32_t rc = TILEDB_OK
         cdef tiledb_ctx_t* ctx_ptr = ctx.ptr
-        # uri
-        cdef bytes buri = unicode_path(uri)
-        cdef const char* uri_ptr = PyBytes_AS_STRING(buri)
-        # mode
-        cdef tiledb_query_type_t query_type = TILEDB_READ
-        # key
-        cdef bytes bkey
-        cdef tiledb_encryption_type_t key_type = TILEDB_NO_ENCRYPTION
-        cdef void* key_ptr = NULL
-        cdef unsigned int key_len = 0
-        # convert python mode string to a query type
-        if mode == 'r':
-            query_type = TILEDB_READ
-        elif mode == 'w':
-            query_type = TILEDB_WRITE
-        else:
-            raise ValueError("TileDB array mode must be 'r' or 'w'")
-        # check the key, and convert the key to bytes
-        if key is not None:
-            if isinstance(key, str):
-                bkey = key.encode('ascii')
+        cdef tiledb_array_schema_t* schema_ptr
+        cdef tiledb_array_type_t array_type
+        cdef Array new_array
+        cdef object new_array_typed
+
+        # *** this call frees array_ptr and raises a Python exception ***
```

Comment:
Can you explain this in a little more detail? I'm not sure where the exception is coming from and where it is caught.

Reply:
Nice catch, thanks. This comment is outdated. Initially, this function was returning a bare `tiledb_array_t*`, but Cython requires [extra care](https://cython.readthedocs.io/en/latest/src/userguide/language_basics.html#error-return-values) when writing functions which do not return a Python object (PyObject). So, to make the `preload` function safer, I added the `ArrayPtr` wrapper class. Any exceptions inside the `preload` function should be properly raised, without needing to worry about freeing `array_ptr`. (after this call returns, the `try..except` block is responsible for freeing `array_ptr` if there are any exceptions while retrieving the schema info)

I will update the comment to reflect the above discussion.